File: backend/miniapps/files/tools/fspath.py
from typing import Tuple
from uuid import UUID

# Valid formats:
# storage + abs path: `storage_id:/path/to/file`
# abs path: `/path/to/file`
# rel path: `path/to/file`
# all paths may end in `/`
SEP = "/"
STORAGE_SEP = ":"
CURRENT_DIR = "."
PARENT_DIR = ".."


class AbsolutePathExpected(Exception):
    def __init__(self, path: str):
        super().__init__(f"Absolute path expected, but got '{path}'")

def isvalid(path: str):
    """Returns True if the path is valid"""
    if not path:
        return False
    storage_id, path = strip_storage(path)
    if storage_id is not None:
        return path.startswith(SEP)
    return True

def hasstorage(path: str):
    """Returns True if the path has a storage ID"""
    return path.find(":") != -1

def strip_storage(path: str) -> Tuple[UUID|str|None, str]:
    """Returns the storage ID and the path without the storage ID"""
    if hasstorage(path):
        colon_idx = path.find(":")
        storage_id = path[:colon_idx]
        path = path[colon_idx + 1:]
        try:
            return UUID(storage_id), path
        except ValueError:
            return storage_id, path
    return None, path

def isabs(path: str):
    """Returns True if the path is absolute"""
    return hasstorage(path) or path.startswith(SEP)

def expect_abs(path: str):
    """Raises an exception if the path is not absolute"""
    if not isabs(path):
        raise AbsolutePathExpected(path)

def join(storage_id: UUID|str|None, *parts: str):
    """Join two or more path components, inserting '/' as needed"""
    if not parts:
        if storage_id is not None:
            return f"{storage_id}{STORAGE_SEP}{SEP}"
        return SEP
    parts = tuple(part.removesuffix(SEP) for part in parts)
    if storage_id is not None:
        if not parts[0].startswith(SEP):
            parts = (SEP + parts[0], *parts[1:])
        return f"{storage_id}{STORAGE_SEP}{SEP.join(parts)}"
    return SEP.join(parts)

def normpath(path: str):
    """Normalize a pathname by collapsing redundant separators and up-level references"""
    storage_id, parts = get_parts(path)
    i = 0
    while i < len(parts):
        parts[i] = parts[i].strip()
        if parts[i] == CURRENT_DIR:
            del parts[i]
        elif parts[i] == PARENT_DIR:
            if i == 0:
                raise AbsolutePathExpected(path)
            del parts[i]
            del parts[i - 1]
            i -= 1
        else:
            i += 1
    return join(storage_id, *parts)

def get_parts(path: str):
    """Returns the storage ID and a list of the path components"""
    storage_id, path = strip_storage(path)
    parts = [part.strip() for part in path.split(SEP) if part]
    return storage_id, parts

def dirname(path: str):
    """Returns the same path, but without the last component"""
    sep_idx = path.rfind(SEP)
    if sep_idx == -1 or sep_idx == len(path) - 1:
        return path
    return path[:sep_idx + 1].strip()

def basename(path: str):
    """Returns the final component of a pathname"""
    sep_idx = path.rfind(SEP)
    if sep_idx == -1:
        _, path = strip_storage(path)
        return path
    return path[sep_idx + 1:].strip()

def ext(path: str):
    """Returns the extension of the path (from the last '.' to end of the string)"""
    period_idx = path.rfind(".")
    if period_idx == -1:
        return ""
    return path[period_idx:].strip()

File: frontend/src/views/files/fspath.ts
import { generatePath } from 'react-router-dom';

const sep = "/";
const storageSep = ":";
const currentDir = ".";
const parentDir = "..";

function isValid(path: string) {
    if (path === "") {
        return false;
    }
    const [storageId, filePath] = stripStorage(path);
    if (storageId !== null) {
        return filePath.startsWith(sep);
    }
    return true;
}

function hasStorage(path: string) {
    return path.indexOf(":") !== -1;
}

function stripStorage(path: string): [string|null, string] {
    if (hasStorage(path)) {
        const colonidx = path.indexOf(":");
        const storageId = path.slice(0, colonidx);
        const filePath = path.slice(colonidx + 1);
        return [storageId, filePath];
    }
    return [null, path];
}

function isAbs(path: string) {
    const colonidx = path.indexOf(":");
    return colonidx !== -1 || path.startsWith(sep);
}

function join(storageId: string|null, parts: string[]) {
    if (parts.length === 0) {
        if (storageId !== null) {
            return `${storageId}${storageSep}${sep}`;
        }
        return sep;
    }
    parts = parts.map((p) => p.endsWith(sep) ? p.slice(0, p.length - 1) : p);
    if (storageId !== null) {
        if (!parts[0].startsWith(sep)) {
            parts = [sep + parts[0], ...parts.slice(1)];
        }
        return `${storageId}${storageSep}${parts.join(sep)}`;
    }
    return parts.join(sep);
}

function normPath(path: string) {
    var [storageId, parts] = getParts(path);
    var i = 0;
    while (i < parts.length) {
        parts[i] = parts[i].trim();
        if (parts[i] == currentDir) {
            parts = parts.splice(i, 1);
        } else if (parts[i] == parentDir) {
            if (i === 0) {
                throw "Expected absolute path";
            }
            parts = parts.splice(i - 1, 2);
            i -= 1;
        } else {
            i += 1;
        }
    }
    return join(storageId, parts)
}

function getParts(path: string): [string|null, string[]] {
    const [storageId, filePath] = stripStorage(path);
    const parts = filePath.split(sep).filter((part) => part !== "");;
    return [storageId, parts];
}

function dirName(path: string) {
    const sepIdx = path.lastIndexOf(sep, path.length - 2);
    if (sepIdx === -1) {
        return path;
    }
    return path.slice(0, sepIdx + 1);
}

function baseName(path: string) {
    const sepIdx = path.lastIndexOf(sep);
    if (sepIdx === -1) {
        const [_, filePath] = stripStorage(path);
        return filePath;
    }
    return path.slice(sepIdx + 1);
}

function ext(path: string) {
    const periodIdx = path.lastIndexOf(".");
    if (periodIdx === -1) {
        return "";
    }
    return path.slice(periodIdx);
}

// Path <-> URL

function pathToUrl(route: string, path: string) {
    var [storageId, filePath] = stripStorage(path);
    if (filePath.startsWith(sep)) {
        filePath = filePath.slice(1);
    }
    return generatePath(route, {
        storageId: storageId,
        "*": filePath,
    });
}

function urlToPath(route: string, url: string) {
    const routeStorageIdx = route.indexOf(":storageId");
    if (route.slice(routeStorageIdx) !== ":storageId/*") {
        throw `File route '${route}' doesn't end in ':storageId/*'`;
    }
    const pathInUrl = url.slice(routeStorageIdx + ":storageId".length + 1);
    const slashIdx = pathInUrl.indexOf("/");
    const storageId = pathInUrl.slice(0, slashIdx);
    const filePath = pathInUrl.slice(slashIdx);
    return join(storageId, [filePath]);
}

export default {
    sep,
    storageSep,
    isValid,
    hasStorage,
    stripStorage,
    isAbs,
    join,
    normPath,
    getParts,
    dirName,
    baseName,
    ext,
    pathToUrl,
    urlToPath,
};
